### DISCUSSION TITLE
Suspend while SSR (nextjs with async atom)

### DISCUSSION THREAD

**Discussion by dnbkr** [pmndrs/jotai#1107] [Q&A]
Hi!

I just started using jotai and suspense now that React 18 is out. I have a question about how to achieve something (or the best way to solve it anyway)

I want to `fetch` some data when my page loads, and then have the result globally available. I'm using NextJS, but I will be exporting a static site, so I don't want the `fetch` to happen on the server render, only on the client.

my first attempt at this was a more traditional fetch-then-store approach:

```typescript
const result = atom<Data | undefined>(undefined);

function useData() {
  const [state, setState] = useAtom(result);
  useEffect(() => fetchData().then(setState), []);
  return state;
}
```

but this doesn't work with suspense, and doesn't seem to be very 'jotai' style? So I looked at using an async atom:

```typescript
const result = atom(async () => await fetchData())

function useData() {
  return useAtomValue(result)
}
```

this works with suspense, and its much nicer IMO but it will resolve the `fetchData` promise on the server side (and the client side I think)

so based on the docs, I got around this using:

```typescript
const result = atom(async () => {
  if (typeof window === 'undefined') {
    return undefined;
  }
  return await fetchData();
})

function useData() {
  return useAtomValue(result)
}
```

and this is great - I doesn't resolve on the server and works with suspense; but the type signature for `useData` is that it returns `Data | undefined` and while client side the result will never be `undefined` because of suspense, on the server side render it will be.

So what I'm after is a way to suspend the async atom while rendering on the server. Something like:

```typescript
const result = atom(async () => {
  if (typeof window === 'undefined') {
    return continueSuspending();
  }
  return await fetchData();
})

function useData() {
  return useAtomValue(result)
}
```

what's the best way to achieve something like this? Is it possible? Maybe it's not a Jotai thing, but more a Next or React thing? It strikes me that my needs here are surely not very niche? Anyone that produces a static site using Next (or, I think, Gatsby? Or any static site generator that does the first render at build time) will have this problem?

Any help is greatly appreciated!

**Comment by dai-shi** [comment]
Nice try. To my knowledge this is still an open space for research.
If there's any best practice with pure React without jotai, we can work on possible jotai style.

Please keep sharing your experiments and hope someone can jump in too.

**Reply by dnbkr** [reply]
thank you for your quick response - if I find a better approach or compromise I'll post here

**Reply by phuctm97** [reply]
Following this because I'm looking to do the exact same thing and no luck suspending on server (with Next.js).

**Comment by maikeriva** [comment]
>this works with suspense, and its much nicer IMO but it will resolve the fetchData promise on the server side (and the client side I think)

I confirm this seems to be the case. If you have an atom with an async getter in a client component and you use the solution which works with suspense:
```
const result = atom(async () => await fetchData())

function useData() {
  return useAtomValue(result)
}
```
It will be rendered on the server first, then a second time on the client. I am also exploring solutions.

**Comment by n-p-e** [comment]
I think I've encountered this exact same problem. Here's a minimal setup to replicate the issue:

```typescript
// pages/bad.tsx
import { atom, useAtom } from "jotai";
import { Suspense } from "react";

const slowAtom = atom(async (): Promise<string> => {
  return new Promise((resolve) => {
    // This causes Next to report error even if behind <Suspense>
    // if (typeof window === "undefined") {
    //   throw new Error("Doesn't work on server");
    // }
    setTimeout(
      () => resolve(typeof window === "undefined" ? "on server" : "on client"),
      5 * 1000
    );
  });
});

const Content = () => {
  const [value] = useAtom(slowAtom);
  return <p>{value}</p>;
};

export default function BadExample() {
  return (
    <Suspense fallback={<p>loading...</p>}>
      <Content />
    </Suspense>
  );
}
```

When you access this page in browser, first it takes 5s for any content to appear because Next executes the atom getter on the server and waits for the promise to resolve. Once the page loads, it shows "on server". Then after another 5 seconds, you get a "hydration failed" error because the client's initial render shows "on client" instead of "on server".

To fix this problem, we want to skip fetching the data (calling the atom) on the server entirely. The server should return fallback and we only mount `<Content>` on the client. I've come up with this solution:

```typescript
// pages/good.tsx
import { atom, useAtom } from "jotai";
import { ReactNode, Suspense, useEffect, useState } from "react";

// put this in your utils module
export const ClientSuspense: React.FC<{
  fallback?: ReactNode;
  children?: ReactNode;
}> = ({ fallback, children }) => {
  const [showingChildren, setShowingChildren] = useState(false);
  useEffect(() => setShowingChildren(true), []);

  if (showingChildren) {
    return <Suspense fallback={fallback}>{children}</Suspense>;
  }
  return fallback;
};

const slowAtom = atom(async (): Promise<string> => {
  return new Promise((resolve) => {
    if (typeof window === "undefined") {
      throw new Error("This won't be hit at all");
    }
    setTimeout(
      () => resolve(typeof window === "undefined" ? "on server" : "on client"),
      5 * 1000
    );
  });
});

const Content = () => {
  const [value] = useAtom(slowAtom);
  return <p>{value}</p>;
};

export default function GoodExample() {
  return (
    <ClientSuspense fallback={<p>Loading...</p>}>
      <Content />
    </ClientSuspense>
  );
}
```

`<ClientSuspense>` is like `<React.Suspense>`, except it skips rendering children entirely on the server. You can wrap all components that use async atoms with this, so your data fetching logic happens only on the client. To be extra safe, insert the `typeof window === 'undefined'` check in your atoms.
